crates/sui/src/sui_commands.rs

```diff
@@ -429,6 +507,215 @@ impl SuiCommand {
+        if let Some(config) = config.clone() {
+            swarm_builder = swarm_builder.dir(config);
+        }
+
+        let epoch_duration_ms = if let Some(epoch_ms) = epoch_duration_ms {
+            epoch_ms
+        } else {
+            60000
+        };
+        swarm_builder = swarm_builder.with_epoch_duration_ms(epoch_duration_ms);
+    } else {
+        // load from config dir that was passed, or generate a new genesis if there is no config
+        // dir passed and there is no config_dir in the default location
+        // and if a dir exists, then use that one
+        if let Some(config) = config.clone() {
+            swarm_builder = swarm_builder.dir(config);
+        } else if config.is_none() && !sui_config_dir()?.join(SUI_NETWORK_CONFIG).exists() {
+            genesis(None, None, None, false, epoch_duration_ms, None, false).await?;
+            swarm_builder = swarm_builder.dir(sui_config_dir()?);
+        } else {
+            swarm_builder = swarm_builder.dir(sui_config_dir()?);
+        }
+        // Load the config of the Sui authority.
+        let network_config_path = config
+            .clone()
+            .unwrap_or(sui_config_dir()?.join(SUI_NETWORK_CONFIG));
+        let network_config: NetworkConfig =
+            PersistedConfig::read(&network_config_path).map_err(|err| {
+                err.context(format!(
+                    "Cannot open Sui network config file at {:?}",
+                    network_config_path
+                ))
+            })?;
+
+        swarm_builder = swarm_builder.with_network_config(network_config);
+    }
+
+    // the indexer requires to set the fullnode's data ingestion directory
+    // note that this overrides the default configuration that is set when running the genesis
+    // command, which sets data_ingestion_dir to None.
+    if with_indexer {
+        swarm_builder = swarm_builder.with_data_ingestion_dir(data_ingestion_path.clone());
+    }
+
+    let mut fullnode_url = sui_config::node::default_json_rpc_address();
+    fullnode_url.set_port(fullnode_rpc_port);
+
+    if no_full_node {
+        swarm_builder = swarm_builder.with_fullnode_count(0);
+    } else {
+        swarm_builder = swarm_builder
+            .with_fullnode_count(1)
+            .with_fullnode_rpc_addr(fullnode_url);
+    }
+
+    let mut swarm = swarm_builder.build();
+    swarm.launch().await?;
+    // Let nodes connect to one another
+    tokio::time::sleep(tokio::time::Duration::from_secs(2)).await;
+    info!("Cluster started");
+
+    // the indexer requires a full url
+    let fullnode_url = format!("http://{}", fullnode_url);
+    info!("Fullnode URL: {}", fullnode_url);
+    let pg_address = format!("postgres://{pg_user}:{pg_password}@{pg_host}:{pg_port}/{pg_db_name}");
+
+    if no_full_node && with_indexer {
+        tracing::warn!("Cannot start the indexer without a fullnode, so will skip the indexer and graphql if requested");
```

Comment:
I am a little lost on the logic of what can be (or has to be) started with what. Perhaps we could have some kind of matrix explaining it (and the code follow this more principled approach to make sure that we did not miss any corner cases that should be reported as errors/conflicts)?

Reply:
Thanks for raising this point. I will revisit this after strengthening my understanding. 

The message should say `Without a fullnode, the indexer will not ingest any local data and will fallback on some formal snapshot` or something along those lines. IIUC it does not make sense to start an indexer without a fullnode, or at least, the indexer needs to be properly configured to pick up data from a source.

Reply:
Regarding the logic of the conditions under these which services should run, I will add some explanation in the description.